File: passwdgen.pl
#!/usr/bin/perl -w
# Password-generating script, v0.03
use strict;
use Getopt::Std;

sub pick_element (@);
sub array_entropy (@);

my @patterns = ();
my $numtogen = 10;
my %opts;

getopts( "hvbt:n:", \%opts );

if ($opts{"h"}) {
  print << "EOF";
Usage: $0 [-hvb] [-t <template>] [-n <number>]
Options:
  -h: Prints this help message and quits.
  -v: Prints version information.
  -t: Sets password template (default: cvcvcvcdd).
  -n: Sets number of passwords to generate (default 10).
  -b: Calculates number of bits in specified pattern; no passwords generated.
  
Password templates use the following characters:
  c     lowercase consonant
  v     lowercase vowel
  l     lowercase letter
  C     mixed-case consonant
  V     mixed-case vowel
  L     mixed-case letter
  d, D  digit
  p, P  punctuation symbol (from !\@\%^\&-_+=?)
  
EOF
  exit();
}
elsif ($opts{v}) {
  print << "EOF";
Password generator script, version 0.03.
EOF
  exit();
}

if ($opts{"t"}) {
	push @patterns, $opts{"t"};
}

if ($opts{"n"}) {
	if (int($opts{"n"}) > 0) {
        $numtogen = $opts{"n"};
    }
}

if (! @patterns) {
	push @patterns, "cvcvcvcdd";
}

my @LC_CONSONANTS = qw( b c d f g h k l m n p r s t v w x z );
my @LC_VOWELS = qw( a e i o u y );

my @MC_CONSONANTS = @LC_CONSONANTS;
push @MC_CONSONANTS, map { uc $_ } @LC_CONSONANTS;

my @MC_VOWELS = @LC_VOWELS;
push @MC_VOWELS, map { uc $_ } @LC_VOWELS;

my @LC_LETTERS = @LC_CONSONANTS;
push @LC_LETTERS, @LC_VOWELS;

my @MC_LETTERS = @LC_LETTERS;
push @MC_LETTERS, map { uc $_ } @LC_LETTERS;

my @DIGITS = 0 .. 9;
my @PUNCT = qw( ! @ % ^ & - _ + = ? );

my %pools = ( 'c' => \@LC_CONSONANTS,
              'C' => \@MC_CONSONANTS,
              'v' => \@LC_VOWELS,
              'V' => \@MC_VOWELS,
              'l' => \@LC_LETTERS,
              'L' => \@MC_LETTERS,
              'd' => \@DIGITS,
              'D' => \@DIGITS,
              'p' => \@PUNCT,
              'P' => \@PUNCT
              );
              
if ($opts{b}) {
   	# just calculate bits in pattern
   	# no longer assumes each character in pool is distinct
   	foreach my $pattern (@patterns) {
   	    my $valuecount = 1;
   	    my $bits = 0;
   	    
   	    while($pattern =~ /(.)/g) {
   	       # print "@{$pools{$1}}\n";
   	       $valuecount *= scalar @{$pools{$1}};
   	       $bits += array_entropy(@{$pools{$1}});
   	    }
   	    
   	    my $vbits = ( log($valuecount) / log(2) );
   	    print "Pattern: $pattern   Bits: $bits\n";
   	}
   	exit();
}

                
foreach my $pattern (@patterns) {
	# print "$pattern\n";
    for( my $i = 0; $i < $numtogen; $i++ ) {
    	my $temp = $pattern;
    	$temp =~ s/(.)/pick_element(@{$pools{$1}})/ge;
    	print "$temp\n";
    }
}

sub pick_element (@) {
    my @array = @_;
    my $index;
    $index = int( scalar @array * rand() );
    return $array[$index];
}

sub array_entropy (@) {
    my @array = @_;
    my %counts = ();
    foreach my $elt (@array) {
        $counts{$elt}++;
    }
    
    # bits computed by entropy formula S = \sum_i -p_i \log_2 p_i
    # which simplifies to S = \log_2 t - \sum_i c_i/t \log_2 c_i
    # in the case that p_i = c_i/t for all i
    
    # start in natural logs
    my $bits = log(@array);
    
    foreach my $key (keys %counts) {
        #print "$key $counts{$key}\n";
        $bits -= $counts{$key} * log($counts{$key}) / @array;
    }   
    
    # convert to base 2 logs
    $bits /= log(2);
    
    return $bits;
}

File: passwdgen.py
#! /usr/bin/env python
import random, sys
import getopt
import argparse
from math import log

version = '0.06'

def buildusage(name):
	return 'Usage: ' + name + """ [-hvb] [-t <template>] [-n <number>]
Options:
  -h: Prints this help message and quits.
  -v: Prints version information.
  -t: Sets password template (default: cvcpcvccvcpdd).
  -n: Sets number of passwords to generate (default 10).
  -b: Calculates number of bits in specified pattern; no passwords generated.

Password templates use the following characters:
  c     lowercase consonant
  v     lowercase vowel
  l     lowercase letter
  C     mixed-case consonant
  V     mixed-case vowel
  L     mixed-case letter
  d, D  digit
  p, P  punctuation symbol (from !@#%^&()-_+=[]{};,.?)
  h, H  hex digit (h for lowercase a-f, H for uppercase)
	"""

def main(argv = None):
	if argv is None:
		argv = sys.argv

	parser = argparse.ArgumentParser(description='Batch password generator script.')

	parser.add_argument('-t', '--template',
		default='cvcpcvccvcpdd',
		help="""Password template. Recognized characters: cvlCVLdph.""")
	parser.add_argument('-n', '--num', type=int, default=10, help='number of passwords to generate')
	parser.add_argument('-b', '--bits', action='store_true')

	parser.add_argument('-k', '--keyboard', 
		choices=['uniform', 'u', 'qwerty', 'q', 'dvorak', 'd'],
		default='uniform',
		help='Keyboard key distribution to prefer: dvorak, qwerty, or uniform')
		
	parser.add_argument('-v', '--version', action='version', 
						version='%(prog)s' + version)
						
	args = parser.parse_args(argv[1:])
	# print args
	
	bitsonly = args.bits
	patterns = []
	patterns.append(args.template)
	numtogen = args.num
	keyboard = args.keyboard

	pools = make_pools(keyboard)

	if len(patterns) == 0:
		patterns.append("cvcpcvccvcpdd")
		
	if bitsonly:
		for pattern in patterns:
			print(pattern + ": " + str(pattern_entropy(pattern, pools)))
	else:
		random.seed()
		for pattern in patterns:
			for i in range(numtogen):
				print(make_passwd(pattern, pools))
				
	return 0
	

def make_passwd(pattern, pools):
	passwd = ""
	for char in pattern:
		passwd += random.choice(pools[char])
	return passwd


def make_pools(keyboard):
	consonants = {
		'u': 'bcdfghjklmnpqrstvwxz',
		'd': (('htns' * 2 + 'd') * 3 + 
			  ('gcrl' * 2 + 'f') * 2 +
			  ('mwvz' * 2 + 'b')) +
			 ('p' * 2 * 2 + ('qjk' * 2 + 'x')),
		'q': 'sdfghjkl' * 4 + 'qwrtp' * 2 + 'zxcvbnm'
	}
	
	consonants['uniform'] = consonants['u']
	consonants['dvorak'] = consonants['d']
	consonants['qwerty'] = consonants['q']
	
	vowels = {
		'u': 'aeiouy',
		'd': ('aoeu' * 2 + 'i') * 2 + 'y',
		'q': 'a' * 3 + 'eyuio' * 2
	}
	
	vowels['uniform'] = vowels['u']
	vowels['dvorak'] = vowels['d']
	vowels['qwerty'] = vowels['q']
	
	punct = {
		'u': '!@#%^&()-_+=[]{};,.?',
		'd': '---,,..//=[]' * 2 + '___??++~!@#$%^&(){}',
		'q': ';;;,./[[]]-=' * 2 + '???{{}}~!@#$%^&()_+'
	}

	punct['uniform'] = punct['u']
	punct['dvorak'] = punct['d']
	punct['qwerty'] = punct['q']
	
	pools = {}
	
	pools['c'] = list(consonants[keyboard])
	pools['C'] = pools['c'] + [ch.upper() for ch in pools['c']]
	
	pools['v'] = list(vowels[keyboard])
	pools['V'] = pools['v'] + [ch.upper() for ch in pools['v']]
	
	pools['l'] = pools['c'] + pools['v']
	pools['L'] = pools['l'] + [ch.upper() for ch in pools['l']]
	
	pools['d'] = [str(digit) for digit in range(10)]
	pools['D'] = pools['d']
	
	pools['p'] = list(punct[keyboard])
	pools['P'] = pools['p']
	
	pools['h'] = pools['d'] + list('abcdef')
	pools['H'] = pools['d'] + list('ABCDEF')

	pools['-'] = ['-']
	
	return pools
	
def array_entropy(seq):
	if len(seq) == 0:
		return 0

	counts = {}
	for elt in seq:
		counts[elt] = 0
	for elt in seq:
		counts[elt] += 1
		
	# compute entropy S = \sum_i -p_i \log_2 p_i
	# if each p_i = c_i/t, reduces to S = \log_2 t - \sum_i c_i/t \log_2 c_i
	bits = log(len(seq))
	
	for key in list(counts.keys()):
		bits -= counts[key] * log(counts[key]) / len(seq)
		
	# convert to base-2 logs
	bits /= log(2)
	
	return bits
	
def pattern_entropy(pattern, pools):
	return sum([array_entropy(pools[ch]) for ch in pattern])
	
if __name__ == "__main__":
	sys.exit(main())
